api/controllers/AssignmentController.js

```diff
@@ -79,54 +78,73 @@ module.exports = class AssignmentController extends classMixin(OrderedDataContro
       private: parameterizedAssignment.dataValues.auxPrivate
     }
 
+    return this.processSolution({ req, res, aux, assignment, solution, parameterizedAssignment })
+  }
+
+  async evaluateSolution (params, correct, incorrect) {
+    const { req, aux, assignment, solution, parameterizedAssignment } = params
     let judgement = { grade: 0, msg: 'Wrong! Please try again.' }
+    const judge = new assignment.Judge(assignment.judge, req.user)
 
     try {
       judgement = await judge.judge(aux, req.user, assignment.Judge, solution)
-      const [solutionModel] = await Solution.findOrCreate(
-        {
-          where: { studentId: req.user.id, parameterizedAssignmentId: paramId },
-          defaults: {
-            studentId: req.user.id,
-            parameterizedAssignmentId: paramId
-          }
-        }
-      )
-
-      if (!parameterizedAssignment.dataValues.solved) {
-        await solutionModel.update({ data: solution })
-      }
+      await this.updateSolution(req, parameterizedAssignment, solution)
+      await this.updateSolved(judgement, parameterizedAssignment)
+      correct(judgement)
     } catch (e) {
-      logger.error(`${e.constructor.name}: ${e.message}`)
-      const errorType = e.constructor.name
-
-      if (
-        errorType === 'CompilationError' ||
-        errorType === 'AssertionError' ||
-        errorType === 'MissingMethodError'
-      ) {
-        return res.status(200).send(
-          {
-            code: 200, judgement: { grade: 0, msg: e.message }
-          }
-        )
+      logger.error(e)
+      if (e instanceof Sequelize.Error) {
+        throw new HTTPError(500, `Server error: ${e.message}`)
       }
 
-      return res.status(500).send(
-        {
-          error: { code: 500, message: e.message }
-        }
-      )
+      incorrect(e)
     }
+  }
 
-    if (judgement.grade > 0 && !parameterizedAssignment.dataValues.solved) {
-      await parameterizedAssignment.update({ solved: true })
+  async processSolution (params) {
+    if (params.assignment.Judge.isAsync) {
+      const assignmentDescription = { name: params.assignment.metadata.name, title: params.assignment.metadata.title }
+      params.res.status(202).send({ code: 202, judgement: { grade: 0, msg: 'Solution is being processed. Please wait!' } })
+
+      await this.evaluateSolution(
+        params,
+        judgement => {
+          appEmitterBus.emit('solution-judgement-available', { ...judgement, assignment: assignmentDescription })
+        },
+        err => {
+          let judgement = { grade: 0, msg: err.message }
+          appEmitterBus.emit('solution-judgement-available', { ...judgement, assignment: assignmentDescription })
+        })
```

Comment:
You can use try/catch here and in the next section instead of two callbacks: When `await` returns, the judgement is available; when there's an error, it throws.

Then at the async usage, do not `await` it, but use `then` and `catch`. In the `sync` case you can try/await it.